File: 503-next-greater-element-ii/next-greater-element-ii.js
var nextGreaterElements = function(nums) {
    let result = new Array(nums.length).fill(-1)
    let stack = []
    let length = nums.length

    for (let i = (length * 2) - 1; i >= 0; i--) {
        let char = nums[i % length]
        if (!stack.length) {
            stack.push(char)
            continue
        }
        while (stack.length && char >= stack.at(-1)) {
            stack.pop()
        }
        if (stack.length) {
            result[i % length] = stack.at(-1)
        }
        stack.push(char)
    }   
    return result

};

File: 503-next-greater-element-ii/next-greater-element-ii.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        output = [-1 for _ in range(len(nums))]
        size = len(nums)
        j = size * 2
        stack = []

        while (j >= 0):
            if (not stack):
                stack.append(nums[j % size])
            else:
                while (stack and nums[j % size] >= stack[-1]):
                    stack.pop()

                if (stack):
                    output[j % size] = stack[-1]
                
                stack.append(nums[j % size])

            j -= 1
            
        return output